disk/disk.c

```diff
@@ -0,0 +1,272 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include "disk.h"
+#include "disk_impl.h"
+#include "paths.h"
+#include "utils/tsh.h"
+#include "utils/fs.h"
+#include "logserver/logserver.h"
+
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <linux/limits.h>
+
+#define MODULE_NAME "disk"
+#define pv_log(level, msg, ...) vlog(MODULE_NAME, level, msg, ##__VA_ARGS__)
+#include "log.h"
+
+#define DISK_ZRAM_CONFIG_PATH "/sys/devices/virtual/block/%s/%s"
+
+static void pv_disk_free(struct pv_disk *disk)
+{
+	if (disk->name)
+		free(disk->name);
+	if (disk->path)
+		free(disk->path);
+	if (disk->uuid)
+		free(disk->uuid);
+	if (disk->mount_ops)
+		free(disk->mount_ops);
+	if (disk->format_ops)
+		free(disk->format_ops);
+	if (disk->provision_ops)
+		free(disk->provision_ops);
+	if (disk->provision)
+		free(disk->provision);
+
+	free(disk);
+}
+
+void pv_disk_empty(struct pv_state *s)
+{
+	if (!s)
+		return;
+
+	int num_disk = 0;
+	struct pv_disk *d, *tmp;
+	struct dl_list *disks = &s->disks;
+
+	if (!disks)
+		return;
+
+	// Iterate over all disks from state
+	dl_list_for_each_safe(d, tmp, disks, struct pv_disk, list)
+	{
+		pv_log(DEBUG, "removing disk %s", d->name);
+		dl_list_del(&d->list);
+		pv_disk_free(d);
+		num_disk++;
+	}
+
+	pv_log(INFO, "removed %d disks", num_disk);
+}
+
+static int mount_disk(struct pv_disk *disk, struct pv_disk_impl *impl)
+{
+	pv_log(DEBUG, "trying to mount disk %s", disk->name);
+
+	pv_disk_status_t status = impl->status(disk);
+	if (status == DISK_STATUS_MOUNTED) {
+		pv_log(DEBUG, "disk %s already mounted", disk->path);
+		return 0;
+	} else if (status == DISK_STATUS_ERROR) {
+		pv_log(WARN, "disk %s status error", disk->path);
+		return -1;
+	}
+
+	pv_log(DEBUG, "disk %s not mounted", disk->name);
+
+	if (impl->format(disk) != 0) {
+		pv_log(WARN, "cannot format %s, the disk will not be used",
+		       disk->path);
+		return -1;
+	}
+
+	pv_log(DEBUG, "disk %s formated", disk->name);
+
+	if (impl->mount(disk) != 0) {
+		pv_log(WARN, "cannot mount %s, the disk will not be used",
+		       disk->path);
+		return -1;
+	}
+
+	pv_log(DEBUG, "disk %s mounted!", disk->name);
+	return 0;
+}
+
+static int umount_disk(struct pv_disk *disk, struct pv_disk_impl *impl)
+{
+	pv_disk_status_t status = impl->status(disk);
+
+	if (status == DISK_STATUS_NOT_MOUNTED) {
+		pv_log(DEBUG, "disk %s already not mounted", disk->path);
+		return 0;
+	} else if (status == DISK_STATUS_ERROR) {
+		pv_log(WARN, "disk %s status error", disk->path);
+		return -1;
+	}
+
+	if (impl->umount(disk) != 0) {
+		pv_log(WARN, "cannot umount %s", disk->path);
+		return -1;
+	}
+
+	return 0;
+}
+
+int pv_disk_mount_handler(struct pv_disk *disk, const char *action)
```

Comment:
I guess we could change the name of this function to pv_disk_mount or something more fit to what it does now

Reply:
This was reimplemented to improve the name :) and avoid passing the action as string. See https://github.com/pantavisor/pantavisor/pull/363